docs/spec_draft.md

```diff
@@ -522,6 +523,83 @@ it only exists to establish data dependencies from `result` to `inputs`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_reduce
+
+### Semantics
+
+Within each process group in the StableHLO grid, performs reduction, using
+`computations`, over the  values of the `operand` tensor from each process and
+produces a `result` tensor.
+
+The operation splits the StableHLO grid into process groups as follows:
+  * `channel_id <= 0` and `use_global_device_ids = false`,
+    `cross_replica(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups).
+
+Afterwards, within each `process_group`:
+  * `operands@receiver = [operand@sender for sender in process_group]` for all
+    `receiver` in `process_group`.
+  * ```
+    result@process[i0, i1, ..., iR-1 =
+        reduce(
+          inputs=[operands@process[j][i0, i1, ..., iR-1] for j in range(size(operands@process))],
+          init_values=[I, such that `computation` and `I` form a monoid],
+          dimensions=[0],
+          body=computation
+        )
+    ```
+
+### Inputs
+
+| Name                    | Type                                         |
+|-------------------------|----------------------------------------------|
+| `operand`               | tensor of any supported type                 |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64` |
+| `channel_id`            | constant of type `si64`                      |
+| `use_global_device_ids` | constant of type `boolean`                   |
+| `computation`           | `function`                                   |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) size(`replica_groups`) $\gt$ 0.
+  * (C2) All values in `replica_groups` are unique.
+  * (C3) `size(replica_groups)` = `num_replicas`.
```

Comment:
`sum(size(replica_groups))` = `num_replicas`

Reply:
I think size means the number of elements of a tensor.

Reply:
ah, yes, I had a list of list in my mind. from the parallel execution semantics section.